### PR TITLE
[json] faster, lighter JSON enconding/decoding

### PR DESCRIPTION
**Pull request by truthbk**
### What does this PR do?

It introduces jsoniter, a faster, leaner json encoder/decoder alternative (https://github.com/json-iterator/go). After exploring a few options decided to go with this alternative (as opposed to `easyjson` or `ffjson`) due to its simple drop-in replacement nature and great numbers compared to some of the other projects. It is still actively maintained. Most of our heavy-lifting is in encoding so the benefits are not going to be as massive as they'd be with a more heavily decoding application, that said the number of allocations should be reduced helping our footprint and garbage collection. 

### Motivation

Just trying to improve agent performance in general, starting with low hanging fruit like JSON encoding. 

### Additional Notes

To highlight the benefits from making the transition to the optimized `jsoniter` lib for our json processing, we kept track of agent process RSS for a few a different nodes that were suffering from high memory usage due to JSON parsing. We built a custom image off our latest stable including these changes, resulting in some nice gains. See attached graphs:

screen shot 2018-08-27 at 11 32 15 am

screen shot 2018-08-27 at 11 32 22 am

You guys can see the clear improvement after making the deployment on Friday August 3rd. Naturally, upon additional profiling, JSON processing was still responsible for a large chunk of the memory in use, but the gains probably warrant the merging of this PR. The new json library is also pretty much a drop-in replacement for the standard `encoding/json` so it shouldn't cause too many headaches to our developers. 

Finally, since `json-iter` doesn't sort the keys, there's also a small improvement in CPU utilization. Key sorting for JSON is not something particularly useful for us, so it was a nice side-effect.

screen shot 2018-08-27 at 11 40 35 am

### CODE REVIEW

**Comment by truthbk** [comment]
Some actual relevant gains seen in our prod environment, so this is probably worth it. Rebasing and removing the `hold` tag. If possible, I'd like to merge this soon, so we have a nice long run before the `6.5.0` RCs are built.

File: release.json [hunk 1]
```diff
@@ -8,6 +8,15 @@
 		"JMXFETCH_VERSION": "0.20.1",
 		"JMXFETCH_HASH": "076dc742d158e888bfff914e022bd1e640bde2b27735e27ed47799dd89058752"
     },
+    "6.4.1-jsonoptimize": {
```
Comment (arbll) [inline release.json]: Let's remove before merging

**Review by arbll — DISMISSED** [review]
Looks good on my end, just a small question:
- What's the story behind https://github.com/DataDog/datadog-agent/commit/0bae8ceceb5370e96ab3c3e68f7188dab9ad34d2 ?

If this has not been done yet, we'll need to heads-up process (and maybe every agent team) as they probably share some code impacted by this PR.

**Comment by truthbk** [comment]
We can give them the heads-up, that's fine. But trust me, they'll know when they have to rebase every single PR in progress 😊

**Review by olivielpeau — DISMISSED** [review]
LGTM, nice!

* Could you sum up quickly here the perf improvements you've been able to confirm during your testing?
* Can you merge https://github.com/DataDog/datadog-agent/commit/0bae8ceceb5370e96ab3c3e68f7188dab9ad34d2 as part of a different PR? (unless I'm missing something I don't think it belongs in this PR, although it looks like a good change 👍 ) 
* everyone will notice the change for sure, but let's still post a heads-up to everyone when merging this :)

Other than that feel free to merge 👌

**Comment by truthbk** [comment]
Previously approved, removed commit `0bae8ce`, will merge in a different PR. Merging. This will cause a lot of PRs some pain with the need to rebase. I will make sure I notify people via slack.

**Pull request merged**